gax-httpjson/src/main/java/com/google/api/gax/httpjson/HttpRequestRunnable.java

```diff
@@ -110,10 +112,32 @@ HttpRequest createHttpRequest() throws IOException {
     for (HttpJsonHeaderEnhancer enhancer : getHeaderEnhancers()) {
       enhancer.enhance(httpRequest.getHeaders());
     }
+
+    // Set Content-Length header to avoid 411s.
+    HttpJsonHeaderEnhancer contentLengthHeader = getContentLengthHeader(jsonHttpContent);
```

Comment:
This is a little bit odd place for this. Few line earlier (L113) we already call the actual enhancers, but then, right after that we create one more on the fly just to call it once. Ideally we would want to have this enhancer as a default one, created in `InstantiatingHttpJsonChannelProvider#createChannel()` method (together with the other, static enhancers). The difference between this enhancer and the other ones (returned by `getHeaderEnhancers()` is that the other ones are static (do not depend on the actual body of the request) and this one is dynamic (does depend on the body, specifically on the size of the body).

I would suggest changing the `HttpJsonHeaderEnhancer` interface (it is `@BetaApi`, so should be ok) to accept the actual `HttpRequest` instead of `HttpHeaders` object. 

Then create two enhancer implementations, one, which will be doing static headers (i.e. one single enhancer for all static headers) and one for dynamic headers, and instantiate both of those in `InstantiatingHttpJsonChannelProvider#createChannel()`. 

Then here, just call the enhancers normally in the loop (like on line 112), but pass the `httpRequest` as an argument (instead of `httpRequest.getHeaders()` as it is done now). 

With the above suggestion in place, the HeaderEnhancers get their fair chunk of responsibility: setting headers (bot static and dynamic). Now they seem like be just doing a simple call to `HttpHeadersUtils.setHeader()` which is too little to justify existence of the small "family" of classes (the Enhancers class hierarchy).

Reply:
Removed all of this.